### PR TITLE
Add display_name filter to reviewSystems

### PR DESCRIPTION
**Pull request by fhlavac**
https://issues.redhat.com/browse/RHCLOUD-13177
Added filtering by `display_name` to the new Remediation wizard "Review systems" step.

Systems filtering

@katierik

### CODE REVIEW

**Comment by katierik** [comment]
AWESOME

File: packages/remediations/src/RemediationWizard/RemediationWizard.js [hunk 1]
```diff
@@ -67,7 +77,16 @@ const RemediationWizard = ({
         }
     };
 
+    const fetchHostNames = (systems = []) => {
+        const perChunk = 50;
+        const chunks = splitArray(systems, perChunk);
+        chunks.map(chunk => {
```
Comment (Hyperkid123) [inline packages/remediations/src/RemediationWizard/RemediationWizard.js:83]: No need for `map` you don't consume the return value. `forEach` is good enough.

File: packages/remediations/src/RemediationWizard/RemediationWizard.js [hunk 2]
```diff
@@ -159,7 +180,20 @@ RemediationWizard.propTypes = {
         systems: propTypes.arrayOf(propTypes.string),
         onRemediationCreated: propTypes.func
     }).isRequired,
-    basePath: propTypes.string
+    basePath: propTypes.string,
+    registry: propTypes.object.isRequired
```
Comment (Hyperkid123) [inline packages/remediations/src/RemediationWizard/RemediationWizard.js:184]: ReducerRegistry is not `object` but class. Use `PropTypes.instanceOf(ReducerRegistry)`

**Review by Hyperkid123 — COMMENTED** [review]
I'd like to see some additional tests. Functions `fetchSystemsInfo` have some pretty complex conditions.

**Comment by fhlavac** [comment]
@Hyperkid123 sure, I'll add the tests in a while, thank you

**Comment by fhlavac** [comment]
@Hyperkid123 enhanced utils.js tests, all complex logic should be covered

**Comment by Hyperkid123** [comment]
@fhlavac should or is?

**Comment by Hyperkid123** [comment]
This bad boy definitely deserves some attention. It should be easy enough to test. Just mount the step and spy on the `fetchSystemsInfo` and its arguments.

```jsx
const playbookSystems = formValues && formValues[EXISTING_PLAYBOOK] && uniqWith(formValues[EXISTING_PLAYBOOK].issues?.reduce((acc, curr) => [
        ...acc,
        ...(curr.systems.map(system => ({ id: system.id, display_name: system.display_name })))
    ], []), isEqual) || [];
```

**Comment by fhlavac** [comment]
@Hyperkid123 finally, I decided to move the logic to utils, created a test for it, and also made some more refactoring.

File: packages/remediations/src/api/inventory.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import instance from '@redhat-cloud-services/frontend-components-utilities/interceptors';
+import { HostsApi } from '@redhat-cloud-services/host-inventory-client';
+
+export const INVENTORY_API_BASE = '/api/inventory/v1';
+export const hosts = new HostsApi(undefined, INVENTORY_API_BASE, instance);
+
+export function getHostsById(systems, { page, perPage }) {
+    return hosts.apiHostGetHostById(systems, undefined, perPage, page)
+    .then((response) => response);
```
Comment (karelhala) [inline packages/remediations/src/api/inventory.js:9]: Extra `then` response.

File: packages/remediations/src/helpers/host-helper.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import instance from '@redhat-cloud-services/frontend-components-utilities/interceptors';
```
Comment (karelhala) [inline packages/remediations/src/helpers/host-helper.js:1]: Duplicate file, it's the same as `packages/remediations/src/api/inventory.js`

File: packages/remediations/src/steps/reviewSystems.js [hunk 5]
```diff
@@ -31,23 +34,40 @@ const ReviewSystems = ({ issues, systems, registry, ...props }) => {
...
-    const playbook = formValues && formValues[EXISTING_PLAYBOOK];
+    const allSystems = useRef([]);
+    const [ allSystemsNamed, setAllSystemsNamed ] = useState([]);
 
-    const allNewSystems = dedupeArray(issues.reduce((acc, curr) => [
-        ...acc,
-        ...(curr.systems || [])
-    ], [ ...systems ]));
+    useEffect(() => {
+        const playbookSystems = formValues && formValues[EXISTING_PLAYBOOK_SELECTED]
```
Comment (karelhala) [inline packages/remediations/src/steps/reviewSystems.js:52]: Safe accessor can be used here
```suggestion
        const playbookSystems = formValues?.[EXISTING_PLAYBOOK_SELECTED]
```

File: packages/remediations/src/steps/reviewSystems.js [hunk 6]
```diff
@@ -31,23 +34,40 @@ const ReviewSystems = ({ issues, systems, registry, ...props }) => {
...
+            : [];
+        setAllSystemsNamed(unionWith(playbookSystems, newSystemsNamed, isEqual));
+        const allNewSystems = dedupeArray(issues.reduce((acc, curr) => [
+            ...acc,
+            ...(curr.systems || [])
+        ], [ ...systems ]));
+        allSystems.current = dedupeArray([
+            ...allNewSystems,
+            ...(playbookSystems.map(system => system.id))
+        ]);
```
Comment (karelhala) [inline packages/remediations/src/steps/reviewSystems.js:63]: You can theoretically merge these two together, but you are sacrificing readability at that point. So your call @fhlavac 
```JS
        allSystems.current = dedupeArray([
            ...dedupeArray(issues.reduce((acc, curr) => [
              ...acc,
              ...(curr.systems || [])
            ], [ ...systems ])),
            ...(playbookSystems.map(system => system.id))
        ]);
```

**Review by Hyperkid123 — APPROVED** [review]

File: packages/remediations/src/utils/utils.js [hunk 7]
```diff
@@ -177,18 +179,45 @@ const changeBulkSelect = (state, action) => {
...
 
+export const splitArray = (inputArray, perChunk) => inputArray.reduce((resultArray, item, index) => {
+    const chunkIndex = Math.floor(index / perChunk);
+    if (!resultArray[chunkIndex]) {
+        resultArray[chunkIndex] = [];
+    }
+
+    resultArray[chunkIndex].push(item);
+    return resultArray;
+}, []);
```
Comment (karelhala) [inline packages/remediations/src/utils/utils.js:214]: This is too complex, you can achieve the same result by using `slice`, you are making too much iterations
```JS
export const splitArray = [...new Array(Math.ceil(inputArray.length/perChunk))].map((_item, key) => inputArray.slice(key * perChunk, (key + 1) * perChunk))
```

You can test it on https://jsben.ch/cD5xN, the results are quite impressive. The sollution with `split` is almost twice as fast for medium arrays. For small ones the difference is just a few percentage, but once we enter the large sized arrays the difference is huge (20 % for the one in this PR).

File: packages/remediations/src/helpers/host-helper.js (hunk 4 above)
Comment (fhlavac) [reply packages/remediations/src/helpers/host-helper.js:1]: Oh, good catch, thanks for reporting :-)

File: packages/remediations/src/steps/reviewSystems.js (hunk 6 above)
Comment (fhlavac) [reply packages/remediations/src/steps/reviewSystems.js:63]: @karelhala I think we can change it, the names of the variables make it still understandable, thank you

File: packages/remediations/src/utils/utils.js (hunk 7 above)
Comment (fhlavac) [reply packages/remediations/src/utils/utils.js]: @karelhala wow, really cool! Thank you :)

**Review by karelhala — APPROVED** [review]

**Pull request merged**